aiida/transports/plugins/ssh.py

```diff
@@ -1294,35 +1294,95 @@ def exec_command_wait(self, command, stdin=None, combine_stderr=False, bufsize=-
 
         # I flush and close them anyway; important to call shutdown_write
         # to avoid hangouts
         ssh_stdin.flush()
         ssh_stdin.channel.shutdown_write()
 
+        # Now I get the output
+        stdout_bytes = []
+        stderr_bytes = []
+        # 10kB buffer (note that this should be smaller than the window size of paramiko)
+        # Also, apparently if the data is coming slowly, the read() command will not unlock even for
+        # times much larger than the timeout. Therefore we don't want to have very large buffers otherwise
+        # you risk that a lot of output is sent to both stdout and stderr, and stderr goes beyond the
+        # window size and blocks.
+        # Note that this is different than the bufsize of paramiko.
+        internal_bufsize = 10 * 1024
+
+        # Set a small timeout on the channels, so that if we get data from both
+        # stderr and stdout, and the connection is slow, we interleave the receive and don't hang
+        stdout.channel.settimeout(0.1)
+        stderr.channel.settimeout(0.1)  # Maybe redundant, as this could be the same channel.
+
+        while True:
+            chunk_exists = False
+
+            if stdout.channel.recv_ready():  # True means that the next .read call will at least receive 1 byte
+                chunk_exists = True
+                try:
+                    piece = stdout.read(internal_bufsize)
+                    stdout_bytes.append(piece)
+                except socket.timeout:
+                    # There was a timeout: I continue as there should still be data
+                    pass
+
+            if stderr.channel.recv_stderr_ready():  # True means that the next .read call will at least receive 1 byte
+                chunk_exists = True
+                try:
+                    piece = stderr.read(internal_bufsize)
+                    stderr_bytes.append(piece)
+                except socket.timeout:
+                    # There was a timeout: I continue as there should still be data
+                    pass
+
+            # If chunk_exists, there is data (either already read and put in the std*_bytes lists, or
+            # still in the buffer because of a timeout). I need to loop.
+            # Otherwise, there is no data in the buffers, and I enter this block.
+            if not chunk_exists:
+                # Both channels have no data in the buffer
+                if channel.exit_status_ready():
+                    # The remote execution is over
+
+                    # I think that in some corner cases there might still be some data,
+                    # in case the data arrived between the previous calls and this check.
+                    # So we do a final read. Since the execution is over, I think all data is in the buffers,
+                    # so we can just read the whole buffer without loops
+                    stdout_bytes.append(stdout.read())
+                    stderr_bytes.append(stderr.read())
```

Comment:
Is it possible that (for example on heavy load of the client machine or in a memory pressure situation/when swapping) that the channel timeout is triggered here? It's a small buffer and everything should be local and closed, but might be worth to do another try/except to avoid silent content corruption.

Reply:
Thanks for the review! By a look at paramiko's code, I don't think this should happen; [read()](https://github.com/paramiko/paramiko/blob/5cf145ffd3d7cd4d96a511376267a7f27cce5eed/paramiko/file.py#L172), as you say, should only use the local buffer in memory and not access the `recv` method (called within `self._read` of ChannelFile), that in turn is the one that could raise a timeout error.
Also, I fear that if I try to catch the exception and `pass`, I risk to enter an infinite loop. In the current code at least this function would raise in your scenario, and people will notice and we'll have a bug report.

As a note, at least a few people like @adegomme and @ConradJohnston have been testing this implementation and as far as I know it works fine with them, so I would be tempted not to make changes if we're not 100% sure, to avoid to introduce even more subtle bugs.

Reply:
(I'm resolving this conversation but please reopen if you instead believe I should change the code, I'm not 100% sure how, though)